Objects/descrobject.c

```diff
@@ -1790,3 +1790,423 @@ PyTypeObject PyProperty_Type = {
+        if (is_typevar(arg)) {
+            Py_ssize_t iparam = tuple_index(alias->parameters, nparams, arg);
+            assert(iparam >= 0);
+            if (is_tuple) {
+                arg = PyTuple_GET_ITEM(item, iparam);
+            }
+            else {
+                assert(iparam == 0);
+                arg = item;
+            }
+        }
+        Py_INCREF(arg);
+        PyTuple_SET_ITEM(newargs, iarg, arg);
+    }
+    PyObject *res = Py_GenericAlias(alias->origin, newargs);
+    Py_DECREF(newargs);
+    return res;
+}
+
+static PyMappingMethods ga_as_mapping = {
+    .mp_subscript = ga_getitem,
+};
+
+static PyObject *
+ga_call(PyObject *self, PyObject *args, PyObject *kwds)
+{
+    gaobject *alias = (gaobject *)self;
+    PyObject *obj = PyObject_Call(alias->origin, args, kwds);
+    if (obj != NULL) {
+        PyObject_SetAttrString(obj, "__orig_class__", self);
+        PyErr_Clear();
+    }
+    return obj;
+}
+
+static const char* const attr_exceptions[] = {
+    "__origin__",
+    "__args__",
+    "__parameters__",
+    "__mro_entries__",
+    "__reduce_ex__",  // needed so we don't look up object.__reduce_ex__
+    "__reduce__",
+    "__setstate__",
+    NULL,
+};
+
+static PyObject *
+ga_getattro(PyObject *self, PyObject *name)
+{
+    gaobject *alias = (gaobject *)self;
+    if (PyUnicode_Check(name)) {
+        for (const char * const *p = attr_exceptions; ; p++) {
+            if (*p == NULL) {
+                return PyObject_GetAttr(alias->origin, name);
+            }
+            if (PyUnicode_CompareWithASCIIString(name, *p) == 0) {
+                break;
+            }
+        }
+    }
+    return PyObject_GenericGetAttr(self, name);
+}
+
+static PyObject *
+ga_richcompare(PyObject *a, PyObject *b, int op)
+{
+    if (Py_TYPE(a) != &Py_GenericAliasType ||
+        Py_TYPE(b) != &Py_GenericAliasType ||
+        (op != Py_EQ && op != Py_NE))
+    {
+        Py_RETURN_NOTIMPLEMENTED;
+    }
+
+    if (op == Py_NE) {
+        PyObject *eq = ga_richcompare(a, b, Py_EQ);
+        if (eq == NULL)
+            return NULL;
+        Py_DECREF(eq);
+        if (eq == Py_True) {
+            Py_RETURN_FALSE;
+        }
+        else {
+            Py_RETURN_TRUE;
+        }
+    }
+
+    gaobject *aa = (gaobject *)a;
+    gaobject *bb = (gaobject *)b;
+    PyObject *eq = PyObject_RichCompare(aa->origin, bb->origin, Py_EQ);
+    if (eq == NULL)
+        return NULL;
+    if (eq == Py_False) {
+        return eq;
+    }
+    Py_DECREF(eq);
+    return PyObject_RichCompare(aa->args, bb->args, Py_EQ);
+}
+
+static PyObject *
+ga_mro_entries(PyObject *self, PyObject *args)
+{
+    gaobject *alias = (gaobject *)self;
+    return PyTuple_Pack(1, alias->origin);
+}
+
+static PyObject *
+ga_instancecheck(PyObject *self, PyObject *Py_UNUSED(ignored))
+{
+    return PyErr_Format(PyExc_TypeError,
+                        "TypeError: Subscripted generics cannot be used with class and instance checks",
```

Comment:
I don't think this should spell out "TypeError: " explicitly, it produces this:
```pycon
>>> isinstance([], list[str])
...
TypeError: TypeError: Subscripted generics cannot be used with class and instance checks
```

I would also suggest this for the message which is consistent with the message for other invalid use:
"isinstance() arg 2 cannot be a parametrized generic"

Reply:
By the way, shouldn't we also disallow:
```pycon
>>> issubclass(list[str], list)
False
```

Reply:
Ah that's on me, I'll submit fixes for these.

Reply:
Okay hm, for the case of:

```
>>> issubclass(list[str], list)
```

I'm stumped. Perhaps this needs to be implemented in the builtin `isinstance`/`issubclass`? Otherwise I don't see how to prevent that pattern.